core/src/main/scala/org/apache/spark/rdd/ZippedPartitionsRDD.scala

```diff
@@ -58,10 +63,14 @@ private[spark] abstract class ZippedPartitionsBaseRDD[V: ClassTag](
         s"Can't zip RDDs with unequal numbers of partitions: ${rdds.map(_.partitions.length)}")
     }
     Array.tabulate[Partition](numParts) { i =>
-      val prefs = rdds.map(rdd => rdd.preferredLocations(rdd.partitions(i)))
-      // Check whether there are any hosts that match all RDDs; otherwise return the union
-      val exactMatchLocations = prefs.reduce((x, y) => x.intersect(y))
-      val locs = if (!exactMatchLocations.isEmpty) exactMatchLocations else prefs.flatten.distinct
+      val locs = if (preferredLocationRDD != null) {
+        preferredLocationRDD.preferredLocations(preferredLocationRDD.partitions(i))
```

Comment:
What happens when there are no preferred location for a partition in preferredLocationRDD ?
Do we want to fallback to earlier computation ? (which becomes a distinct union I guess)